### ISSUE TITLE
Setup standardized git commit template

### ISSUE THREAD

**Issue by kloudkl** [BVLC/caffe#1181]
Caffe has seen six hundred forks as of creating this issue. The contributors will no doubt continue to grow and diversify. For they and especially the reviewers to understand each others commits easily even with little knowledge of the initial issues, the commit messages should be much more comprehensive than one-liner commits with empty or almost empty message bodies. 

It might be overkill to follow the heavy weight [git commit messages guideline of OpenStack](https://wiki.openstack.org/wiki/GitCommitMessages) which serves to coordinate thousands of contributors from hundreds of organizations. Instead, [setting up a relatively light weight commit template](http://www.git-scm.com/book/ch7-1.html) would be good balance between the two extreme cases. A simple example is [here](https://github.com/netty/netty/wiki/Writing-a-commit-message).

As a consequence of writing more self-contained commit messages, all of the commits of a PR had better be squashed into a single one to keep the one to one relation between commits and PRs and to reduce the writing burden. To automatically create release change logs from the git log, the issue/PR ID should always be set at the end of the commit messages.

**Comment by shelhamer** [comment]
A key part of PR quality is communication. High quality contributions come with thoughtful PR descriptions and commit messages and I'm afraid there's no template for consideration.

I like my commits atomic and echo @longjon's thoughts on the matter. I'm in total disagreement that PRs should be a single commit -- in fact, single commit PRs of any non-trivial change are the most difficult to review.

A PR is ideally one pure thought and the commits carry out the steps to realize it. #594 and other PRs by @longjon and @jeffdonahue are fine examples in this regard.

Closing since we have not come up with a commit template and it is not clear at least to me how it will help. Instead there should be a place in the contributing guide for the details of

> I prefer that each commit be a small, logical, nonbreaking unit. This makes it easier to find bugs (e.g., using git bisect) and easier to understand the history. (As an example, I used nineteen commits in #594, and I think it would be quite hard to follow as a single one!) However, I do support keeping PR history clean by rewriting (e.g., PRs should never have commits to revert changes or fix typos, lint errors, or warnings within the PR; the commits should be rewritten to be right in the first place). The tradeoff is that this puts a higher bar on contributions, since contributors need to be comfortable with rewriting history.

although seeing it all at once might be intimidating. Suggestions to improve the developing and contributing guide are definitely welcome.
